### PR TITLE
cli: add --config-dir common option

### PR DESCRIPTION
**Pull request by Exirel**
I used to left a comment like this:

```
# TODO: allow to use a different homedir
```

But no more. Now, you can use `--config-dir <configdir>` for that. This option provides the location to look for configuration file. Note that this does **not** replace and must **not** be confused with the settings's homedir, which is used internally by Sopel once it is started.

**Edit**: used to be `-H/--homedir <homedir>`

### CODE REVIEW

**Comment by dgw** [comment]
> the settings's

Ahem, config's* (now I see a bit more inside that brain of yours, replacing `config` with `settings` all over the place even though it just makes things more confusing sometimes)

> Note that this does **not** replace and must **not** be confused with

This is a design consideration. Should we not call this a "homedir"? Should this "homedir" override the other one if given? If this is only for finding config files, maybe it should be called `--config-dir` instead…

Just thinking aloud, here. I'm very much in favor of having this option; it's just, as usual, the details. :sweat_smile:

**Comment by Exirel** [comment]
> This is a design consideration. Should we not call this a "homedir"? Should this "homedir" override the other one if given? If this is only for finding config files, maybe it should be called `--config-dir` instead…

Hm. Yeah, probably. It was called "homedir" long before I added the constant `DEFAULT_HOMEDIR` (hence my choice of name), so I kept that. But you're probably right. To be honest, how configuration are found is bonkers, because it's probably the first time I ever see a software asking for the *name* of the configuration, instead of the path.

That beind said, yes, let's go for  `--config-dir`, that sound reasonable—and way better.

**Comment by dgw** [comment]
I mean, we can totally throw the whole "config name" thing in the trash along with the CLI redesign. Ordinarily I'd say that should wait for Sopel 8 and not add any more to 7's plate, but the time to do it would be in 7 with the CLI changes already being made…

Is it useful enough to have a "list configs" function, that we want to keep the way things work now? That is, specifying a config directory and name separately? Or should we nix that and make the new CLI do `-c` = `--config` = `/path/to/config/file.cfg` (or `.yml`… :smirk:)?

**Comment by Exirel** [comment]
Hmmmm yeah. I think I'm on board with that idea.

However, this is heading into the "hijack a PR to talk about an issue" territory, and it's a dangerous place.

**Comment by dgw** [comment]
> However, this is heading into the "hijack a PR to talk about an issue" territory, and it's a dangerous place.

I guess I should at least _try_ to avoid hypocrisy, huh? Though it's relevant because if we choose to go that alternate route, this PR isn't needed and should be closed (or overwritten with that solution). :wink:

**Comment by Exirel** [comment]
> if we choose to go that alternate route, this PR isn't needed and should be closed

Indeed. At the moment, to be honest, I don't have any upgrade path in mind. I wish I already knew how to migrate from `-c name` but I don't. Even if I don't like this, even if I think it's bonkers, I prefer to take a safer, longer path for the user experience's sake: I don't see how not to break the system with such a change.

Heh, I guess sometimes I can't have it all.

**Comment by Exirel** [comment]
@dgw changed to `--config-dir`

**Comment by Exirel** [comment]
After a second thought: support for `-c name` is unrelated to "where are the config file?" For example, we could have `-c name.cfg` and still have `--config-dir /etc/sopel`. From my point of view, I could have a set of configurations in a shared folder, and run multiple instance of Sopel using this kind of command line:

```
export SOPEL_CONFIG_DIR = "/etc/sopel"
sopel -c freenode.cfg --config-dir $SOPEL_CONFIG_DIR
sopel -c quakenet.cfg --config-dir $SOPEL_CONFIG_DIR
sopel -c private.cfg --config-dir $SOPEL_CONFIG_DIR
```

This is exactly the kind of thing I do for other services. So, heh. You can review that, it's not a feature that will disappear when we rework the `-c name` option. ;-)

File: sopel/cli/config.py [hunk 1]
```diff
@@ -22,11 +22,13 @@ def build_parser():
...
     list_parser = subparsers.add_parser(
         'list',
-        help="List available configurations from Sopel's homedir",
+        help="List available configurations from Sopel's config directory",
         description="""
-            List available configurations from Sopel's homedir ({homedir})
-            with the extension "{ext}".
-        """.format(homedir=config.DEFAULT_HOMEDIR, ext='.cfg'))
+            List available configurations from Sopel's config directory
+            with the extension "{ext}". Use option --config-dir to use a
```
Comment (dgw) [inline sopel/cli/config.py]: Should we do this to align with changes in #1646?

```suggestion
            with the extension "{ext}". Use option ``--config-dir`` to use a
```

File: sopel/cli/config.py [hunk 2]
```diff
@@ -75,15 +77,16 @@ def handle_list(options):
     It is possible to filter by extension using the ``-e/--ext/--extension``
```
Comment (dgw) [inline sopel/cli/config.py]: Hmm, spotted another of this pattern I just mentioned in my review of #1646. Not necessary to change it in this PR, but maybe you want to do so in the other. 😸

File: sopel/cli/run.py [hunk 3]
```diff
@@ -131,7 +131,10 @@ def add_legacy_options(parser):
                             "use `sopel restart` instead)"))
     parser.add_argument("-l", '--list', action="store_true",
                         dest="list_configs",
-                        help="List all config files found")
+                        help=(
+                            "List all config files found"
+                            "(deprecated, and will be removed in Sopel 8; "
+                            "use `sopel-config list` instead)"))
```
Comment (dgw) [inline sopel/cli/run.py]: Someone remind me to add this in the v7 migration guide's CLI overhaul section.

File: sopel/cli/utils.py [hunk 4]
```diff
@@ -199,17 +199,21 @@ def add_common_arguments(parser):
     :type parser: argparse.ArgumentParser
 
     This functions adds the common arguments for Sopel's command line tools.
-    At the moment, this functions adds only one argument to the parser: the
-    argument used as the standard way to define a configuration filename.
+    It adds the following arguments:
+
+    * ``-c/--config``: the name of the Sopel config, or its absolute path
```
Comment (dgw) [inline sopel/cli/utils.py]:
```suggestion
    * ``-c``/``--config``: the name of the Sopel config, or its absolute path
```

File: sopel/cli/utils.py [hunk 5]
```diff
@@ -199,17 +199,21 @@ def add_common_arguments(parser):
...
+    * ``--config-dir``: the directory to look for Sopel config
```
Comment (dgw) [inline sopel/cli/utils.py]:
```suggestion
    * ``--config-dir``: the directory to scan for config files
```

or

```suggestion
    * ``--config-dir``: the directory in which to look for config files
```

File: sopel/cli/utils.py [hunk 6]
```diff
@@ -219,10 +223,15 @@ def add_common_arguments(parser):
...
+        default='default',
         metavar='filename',
         dest='config',
-        help='Use a specific configuration file')
+        help='Use a specific configuration file.')
+    parser.add_argument(
+        '--config-dir',
+        default=config.DEFAULT_HOMEDIR,
+        dest='configdir',
+        help='Look for configuration from this Sopel config directory.')
```
Comment (dgw) [inline sopel/cli/utils.py]:
```suggestion
        help='Look for configuration files in this directory.')
```

File: sopel/cli/config.py (hunk 2 above)
Comment (Exirel) [inline sopel/cli/config.py]:
```suggestion
    It is possible to filter by extension using the ``-e``/``--ext``/``--extension``
```

File: sopel/cli/run.py (hunk 3 above)
Comment (Exirel) [reply sopel/cli/run.py]: Ping. :)

**Comment by Exirel** [comment]
@dgw All good! Also:

> Someone remind me to add this in the v7 migration guide's CLI overhaul section.

File: sopel/cli/run.py [hunk 7]
```diff
@@ -291,10 +294,10 @@ def print_version():
     print('https://sopel.chat/')
 
 
-def print_config():
-    """Print list of available configurations from default homedir."""
-    configs = utils.enumerate_configs(config.DEFAULT_HOMEDIR)
-    print('Config files in %s:' % config.DEFAULT_HOMEDIR)
+def print_config(configdir):
+    """Print list of available configurations from default config directory."""
```
Comment (dgw) [inline sopel/cli/run.py]: Here, too, it makes little sense to say "default" any more.

File: sopel/cli/utils.py [hunk 8]
```diff
@@ -266,7 +278,8 @@ def load_settings(options):
     elif 'SOPEL_CONFIG' in os.environ:
         name = os.environ['SOPEL_CONFIG'] or name  # use default if empty
 
-    filename = find_config(config.DEFAULT_HOMEDIR, name)
+    filename = find_config(
+        getattr(options, 'configdir', config.DEFAULT_HOMEDIR), name)
```
Comment (dgw) [inline sopel/cli/utils.py]: Why is this using `getattr`? `options.configdir` has a default value, defined just a few dozen lines earlier.

File: test/cli/test_cli_utils.py [hunk 9]
```diff
@@ -99,14 +100,29 @@ def test_add_common_arguments():
...
+    assert options.configdir == 'test-long-dir'
+
+    options = parser.parse_args(
+        ['--config', 'test-long', '--config-dir', 'test-long-dir'])
+    assert options.config == 'test-long'
+    assert options.configdir == 'test-long-dir'
+
 
 def test_add_common_arguments_subparser():
     """Assert function adds the -c/--config option on a subparser."""
```
Comment (dgw) [inline test/cli/test_cli_utils.py]: Should update this docstring.

File: sopel/cli/utils.py (hunk 8 above)
Comment (Exirel) [reply sopel/cli/utils.py]: Good call!

**Comment by Exirel** [comment]
@dgw Aside from the random error (caused by a test that use an online resource), it's all good to a new review.

File: sopel/cli/run.py (hunk 3 above)
Comment (dgw) [reply sopel/cli/run.py]: Double-backtick note still applies to latest PR revision.

File: test/cli/test_cli_utils.py [hunk 10]
```diff
@@ -104,32 +105,65 @@ def test_add_common_arguments():
...
+        ['--config', 'test-long', '--config-dir', 'test-long-dir'])
+    assert options.config == 'test-long'
+    assert options.configdir == 'test-long-dir'
+
 
 def test_add_common_arguments_subparser():
-    """Assert function adds the -c/--config option on a subparser."""
+    """Assert function adds the multiple options on a subparser.
+
+    The expected options are -c/--config and --config-dir.
```
Comment (dgw) [inline test/cli/test_cli_utils.py]: Backticks for option names?

Comment (Exirel) [reply test/cli/test_cli_utils.py]: In a test, unless we want to generate a doc from them (we could, that can have values), I'm not really for it.

Comment (dgw) [reply test/cli/test_cli_utils.py]: I think it's safer to write docstrings in the same style no matter where they appear, whether that file is included in the generated docs or not _at present_, so we don't have to chase down bad formatting later if/when the docs start pulling in that file.

**Comment by Exirel** [comment]
@dgw I fixed the double-backtick problem. All it needs now is #1686 to be merged so tests pass with Python 2.

**Pull request merged**